In [2]:
RAW_FILE = 'data/raw/ukpopulationestimates18382024.xlsx'

# Notebook 1: LOad and Inspect

## Purpose 
Load the ONS UK Population Estimates file (1838-2024) and understand its structure before any cleaning takes place.

## Source 
Office for National Statistics (ONS) - Mid-Year Population Estimates
Pulished: 26 September 2025
File: ukpopulationestimates18382024.xlsx

In [4]:
import pandas as pd
import openpyxl

RAW_FILE = 'data/raw/ukpopulationestimates18382024.xlsx'

wb = openpyxl.load_workbook(RAW_FILE, read_only=True)
print("All sheets in this file.:")
for i, name in enumerate(wb.sheetnames):
    print(f" {i+1}. {name}")

All sheets in this file.:
 1. Cover sheet
 2. Contents
 3. Notes
 4. Related publications
 5. Table 1
 6. Table 2
 7. Table 3
 8. Table 4
 9. Table 5
 10. Table 6
 11. Table 7
 12. Table 8
 13. Table 9
 14. Table 10
 15. Table 11
 16. Table 12
 17. Table 13
 18. Table 14
 19. Table 15
 20. Table 16
 21. Table 17
 22. Table 18


In [7]:
df_t1 = pd.read_excel(RAW_FILE, sheet_name='Table 1', header=1)
print("Table 1 shape:", df_t1.shape)
print("\nFirst 5 rows:")
print(df_t1.head())
print("\nColumn names:", df_t1.columns.tolist())
print("nData types:")
print(df_t1.dtypes)

Table 1 shape: (156, 3)

First 5 rows:
       Year  Total Population    
0  Mid-2024          69281437 NaN
1  Mid-2023          68526183 NaN
2  Mid-2022          67636134 NaN
3  Mid-2021          66977988 NaN
4  Mid-2020          66739867 NaN

Column names: ['Year', 'Total Population', ' ']
nData types:
Year                 object
Total Population      int64
                    float64
dtype: object


In [9]:
df_t3 = pd.read_excel(RAW_FILE, sheet_name='Table 3', header=1)
print("Table 3 shape:", df_t3.shape)
print("\nFirst 8 rows:")
print(df_t3.iloc[:8,:6])
print("\nSex values in dataset:", df_t3['Sex'].unique())
print("\nAge values (first 15):", df_t3['Age'].head(15).tolist())
print("\nMissing values per column:")
print(df_t3.isnull().sum().sum(), "total missing cells")

Table 3 shape: (276, 56)

First 8 rows:
        Age      Sex  Mid-2024  Mid-2023  Mid-2022  Mid-2021
0  All Ages  Persons  69281437  68526183  67636134  66977988
1         0  Persons    667994    671246    694093    675435
2         1  Persons    690113    714612    690155    702628
3         2  Persons    732532    710627    717216    719170
4         3  Persons    728400    737811    733701    730341
5         4  Persons    755117    753649    743942    749076
6         5  Persons    769993    763165    762741    772705
7         6  Persons    779267    781374    786428    767725

Sex values in dataset: ['Persons' 'Males' 'Females']

Age values (first 15): ['All Ages', 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]

Missing values per column:
0 total missing cells


## Inspection Findings

- **File contains 18 sheets**: Cover, Contents, Notes, Tables 1-18
- **Table 1**: UK total population 1851-2024. Two columns: Year, Total Population
- **Table 3**: UK population by sex + single year of age, 1971-2024
- Shape: 276 rows x 56 columns
- Ages: 0 to 89 (age 90 = top-coded)
- Sex: Persons / Males / Females
- Years: Mid-1971 to Mid-2024 (54 year columns)
- Format: WIDE - each year is a seperate columns
- ** Key challenge**: Must reshape Table 3 from WIDE to LONG format (one row per Age / Sex / Year combination) before analysis

## Deliverable Summary - Notebook 01: Load and Inspect
This notebook successfully performs the initial inspection of the ONS UK Population Estimates dataset (1838-2024). The following tasks were completed:

## 1. File Loading
- Loaded the raw Excel file from 'data/raw/ukpopulationestimates18382024.xlsx'

## 2. Table-Level Inspection
- **Table 1**: UK total population (1838-2024)
- Clean, simple structure
- Suitable for immediate time-series analysis
- ** Table 3**: Population by sex and single year of age (1971-2024)
- Wide formate (one column per year)
- Requires reshaping before analysis

## 4. Next Steps (Notebook 02)
- Clean column names
- Reshape Table 3 using 'pd.melt()'
- Convert year labels to numeric
- Validate ranges and missing values
- Prepare cleaned datasets for analysis and visualization

**Notebook 01 is complete.**